hazelcast/src/main/java/com/hazelcast/internal/util/collection/FixedCapacityArrayList.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2008-2022, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util.collection;
+
+import javax.annotation.concurrent.NotThreadSafe;
+import java.lang.reflect.Array;
+import java.util.AbstractList;
+import java.util.Arrays;
+
+/**
+ * A very fast implementation of {@link java.util.List} that stores its content in a not resizeable array.
+ * The {@link FixedCapacityArrayList} provides an API to fetch that array. Fetching that array makes
+ * the list instance unusable.
+ */
+@NotThreadSafe
+public class FixedCapacityArrayList<E> extends AbstractList<E> {
+    private E[] elementData;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public FixedCapacityArrayList(Class<E> eClass, int initialCapacity) {
+        this.elementData = (E[]) Array.newInstance(eClass, initialCapacity);
+    }
+
+    @Override
+    public boolean add(E e) {
+        modCount++;
+        elementData[size++] = e;
```

Comment:
I'd add some bounds check and workaround in case when `size >= initialCapacity ` :)

Reply:
This is the line that I didn't want to cross. You can either have a performance or usability here. Here is a JMH benchmark with check vs nocheck:

```java
public class FastListBenchmark {
    private static final Object ENTRY = new Object();
    private static final int COUNT = 100;

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    public void noCheck(Blackhole blackhole) {
        List<Object> list = new FixedCapacityArrayList<>(Object.class, COUNT);
        for (int i = 0; i < COUNT; i++) {
            list.add(ENTRY);
        }
        blackhole.consume(list);
    }

    @Benchmark
    @BenchmarkMode(Mode.AverageTime)
    @OutputTimeUnit(TimeUnit.NANOSECONDS)
    public void check(Blackhole blackhole) {
        List<Object> list = new FixedCapacityArrayListChecked<>(Object.class, COUNT);
        for (int i = 0; i < COUNT; i++) {
            list.add(ENTRY);
        }
        blackhole.consume(list);
    }
}

public class FixedCapacityArrayList<E> extends AbstractList<E> {
    private E[] elementData;
    private int size;

    @SuppressWarnings("unchecked")
    public FixedCapacityArrayList(Class<E> eClass, int initialCapacity) {
        this.elementData = (E[]) Array.newInstance(eClass, initialCapacity);
    }

    @Override
    public boolean add(E e) {
        modCount++;
        elementData[size++] = e;
        return true;
    }

    @Override
    public E get(int index) {
        return elementData[index];
    }

    @Override
    public int size() {
        return size;
    }

    /**
     * Returns a storage with all the elements and makes an instance unusable.
     */
    public E[] asArray() {
        E[] result = size == elementData.length ? elementData : Arrays.copyOfRange(elementData, 0, size);
        elementData = null;
        return result;
    }
}

public class FixedCapacityArrayListChecked<E> extends AbstractList<E> {
    private E[] elementData;
    private int size;

    @SuppressWarnings("unchecked")
    public FixedCapacityArrayListChecked(Class<E> eClass, int initialCapacity) {
        this.elementData = (E[]) Array.newInstance(eClass, initialCapacity);
    }

    @Override
    public boolean add(E e) {
        modCount++;
        if (size == elementData.length) {
            return false;
        }
        elementData[size++] = e;
        return true;
    }

    @Override
    public E get(int index) {
        return elementData[index];
    }

    @Override
    public int size() {
        return size;
    }

    /**
     * Returns a storage with all the elements and makes an instance unusable.
     */
    public E[] asArray() {
        E[] result = size == elementData.length ? elementData : Arrays.copyOfRange(elementData, 0, size);
        elementData = null;
        return result;
    }
}
```

and the results:
```
Benchmark                  Mode  Cnt    Score   Error  Units
FastListBenchmark.check    avgt   25  395,450 ± 1,279  ns/op
FastListBenchmark.noCheck  avgt   25  304,380 ± 1,211  ns/op
```

That single check makes the list slower over 30%. The JVM has decent check already that will result in ```IndexOutOfBoundsException``` - I would prefer to stay this that way - simply for a performance

Reply:
Ah, got you. My question mostly focus data structure DX & re-useablility, but you highlight that it's intended for performance only. Gotcha, no problem :)

Reply:
We can do it as an `assert`, that can be disabled in production.

Reply:
Also, if we want to trade performance for safety, we can also stop incrementing the `modCount`